src/main/java/com/amplitude/Retry.java

```diff
@@ -0,0 +1,244 @@
+                                    break;
+                                }
+                            }
+                            if (!shouldRetry) {
+                                break;
+                            }
+                            if (shouldReduceEventCount && !isLastTry) {
+                                eventCountHolder[0] = eventCountHolder[0] / 2;
+                            }
+                            // Clean up the events
+                            eventsBuffer.subList(0, eventCountHolder[0]).clear();
+                            eventsInRetry -= eventCountHolder[0];
+                        } catch (InterruptedException e) {
+                        }
+                    }
+                });
+        retryThread.start();
+    }
+
+    // Call this function if event not in current Retry list.
+    private static void onEventsError(List<Event> events, Response response, String apiKey) {
+        List<Event> eventsToRetry = events;
+        // Filter invalid event out based on the response code.
+        if (response.status == Status.RATELIMIT && response.RateLimitBody != null) {
+            // JSONObject deviceId as key, number as value
+            JSONObject exceededDailyQuotaUsers = response.RateLimitBody.getJSONObject("exceededDailyQuotaUsers");
+            JSONObject exceededDailyQuotaDevices = response.RateLimitBody.getJSONObject("exceededDailyQuotaDevices");
+            eventsToRetry = events.stream()
+                    .filter((event -> !(event.userId != null && exceededDailyQuotaUsers.has(event.userId))
+                            && !(event.deviceId != null && exceededDailyQuotaDevices.has(event.deviceId))))
+                    .collect(Collectors.toList());
+        } else if (response.status == Status.INVALID) {
+            if ((response.InvalidRequestBody.has("missingField") &&
+                    response.InvalidRequestBody.getString("missingField").length() > 0) ||
+                    events.size() == 1) {
+                // Return early if there's an issue with the entire payload
+                // or if there's only one event and its invalid
+                return;
+            } else if (response.InvalidRequestBody != null) {
+                // Filter out invalid events id
+                int[] invalidEventIndices = collectInvalidEventIndices(response);
+                eventsToRetry = IntStream.range(0, events.size())
+                        .filter(i -> Arrays.binarySearch(invalidEventIndices, i) < 0)
+                        .mapToObj(events::get)
+                        .collect(Collectors.toList());
+            }
+        } else if (response.status == Status.SUCCESS) {
+            return;
+        }
+
+        for (Event event : eventsToRetry) {
+            String userId = (event.userId != null) ? event.userId : "";
+            String deviceId = (event.deviceId != null) ? event.deviceId : "";
+            if ((userId.length() > 0) || (deviceId.length() > 0)) {
```

Comment:
Unnecessary parentheses inside the if statement

Reply:
Symbols are evaluated in a special and set (i.e. static and working with static types) way. Very fun stuff, this is just extra reading: https://introcs.cs.princeton.edu/java/11precedence/